Sources/Datadog/Core/System/MobileDevice.swift

```diff
@@ -99,3 +100,34 @@ internal class MobileDevice {
         }
     }
 }
+
+private final class LowPowerModeMonitor {
+    
+    var isLowPowerModeEnabled: Bool {
+        publisher.currentValue
+    }
+    
+    private var publisher: ValuePublisher<Bool>
+    private var powerStateDidChangeObserver: Any?
+    
+    init(_ processInfo: ProcessInfo) {
+        publisher = ValuePublisher(initialValue: processInfo.isLowPowerModeEnabled)
+        
+        powerStateDidChangeObserver = NotificationCenter
+            .default
+            .addObserver(forName: .NSProcessInfoPowerStateDidChange,
+                         object: nil,
+                         queue: .main) { [weak self] notification in
+                guard let processInfo = notification.object as? ProcessInfo else {
+                    return
+                }
+                self?.publisher.publishAsync(processInfo.isLowPowerModeEnabled)
```

Comment:
Hi @earltedly 👋 

Can you help me understand how accessing `processInfo` like this prevent against calling the deadlock?

I'm confused because the object that posts the notification will be the same object that we query for `isLowPowerModeEnabled` which cause the same issue.

Reply:
Hi @yousefhamza 👋 

It's just as @buranmert, at a _worst case_ this will greatly reduce the number of times DatadogSDK access `processInfo` and thus reduces the potential for hitting this bug.

We can also hope that `processInfo` itself only notifies when the state has recently changed and that immediately querying it will be unlikely to cause `processInfo` to immediately find it has changed again (and thus re-broadcast within the same callstack).

Also, it would be great to hear if you're experiencing this too and in what kind of numbers.

Reply:
Here's the radar: FB9661108. Filing a duplicate would be great as it increases the priority within Apple's triage system.

Reply:
@buranmert @earltedly @ncreated  Thanks for the very thorough responses 🙏 

What I understand is it will help with this fix because:
1. It reduces how many times `isLowPowerModeenabled` is queried
1. Querying it after it has changed will give a much lower chance that it will change again while querying
1. Reading it from a single thread will reduce the chances of multiple reads (which can cause firing a notification) at the same time, while making it thread safe.

I'm working on [Instabug SDK](https://instabug.com) and [this issue came into my radar](https://twitter.com/steipete/status/1443506043733032962?s=20) because we use `ProcessInfo.processInfo.isLowPowerModeEnabled`, we are also prone to this crash too. however, we didn't get any reports of it yet.

Reply:
@yousefhamza yes, your reasoning about this change is correct 👍 with the additional note that our `MobileDevice` is a shared component, instantiated only once from SDK init - so we will have only one active subscriber to LPM-change notification.

Thanks for more details, and I hope you will manage to prevent this crash in Instabug SDK 🙂🤞!